### ISSUE TITLE
bug: `fatal error: concurrent map read and map write` introduced since v0.13.7

### ISSUE THREAD

**Issue by TristanSpeakEasy** [pb33f/libopenapi#212]
```
goroutine 8352 [running]:
github.com/pb33f/libopenapi/index.(*SpecIndex).SearchIndexForReferenceByReferenceWithContext(0xc0062e2000, {0x2eaeb88, 0x3a67120}, 0xc0052678b0)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/search_index.go:110 +0x3d0
github.com/pb33f/libopenapi/index.(*SpecIndex).SearchIndexForReferenceByReference(...)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/search_index.go:20
github.com/pb33f/libopenapi/index.(*SpecIndex).SearchIndexForReference(0x9ae79a?, {0xc004192180?, 0xc000ce2910?})
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/search_index.go:28 +0x5c
github.com/pb33f/libopenapi/index.seekRefEnd(0x1878460?, {0xc004192180?, 0xc004192180?})
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:349 +0x1c
github.com/pb33f/libopenapi/index.(*SpecIndex).scanOperationParams(0xc0062e2000, {0xc0015542e0, 0x4, 0x3?}, 0xc004d78f00, {0xc00564db5b, 0x3})
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:367 +0x152
github.com/pb33f/libopenapi/index.(*SpecIndex).GetOperationsParameterCount(0xc0062e2000)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/spec_index.go:1059 +0xb86
github.com/pb33f/libopenapi/index.runIndexFunction.func1(0x7?, 0x1?)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:485 +0x1b
created by github.com/pb33f/libopenapi/index.runIndexFunction in goroutine 1
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:484 +0x32

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0x10?)
        /home/linuxbrew/.linuxbrew/opt/go/libexec/src/runtime/sema.go:62 +0x25
sync.(*WaitGroup).Wait(0xc0032ec698?)
        /home/linuxbrew/.linuxbrew/opt/go/libexec/src/sync/waitgroup.go:116 +0x48
github.com/pb33f/libopenapi/index.(*SpecIndex).BuildIndex(0xc0062e2000)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/spec_index.go:119 +0x1d5
github.com/pb33f/libopenapi/index.createNewIndex(0xc00591df40, 0xc0062e2000, 0x0)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/spec_index.go:95 +0x33e
github.com/pb33f/libopenapi/index.NewSpecIndexWithConfig(0xc00591df40, 0xc003619e80)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/spec_index.go:48 +0x2c5
github.com/pb33f/libopenapi/index.(*Rolodex).IndexTheRolodex(0xc001f41040)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/rolodex.go:315 +0xcc5
github.com/pb33f/libopenapi/datamodel/low/v3.createDocument(0xc006508000, 0xc0054d7ab0)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/datamodel/low/v3/create_document.go:96 +0x5e5
github.com/pb33f/libopenapi/datamodel/low/v3.CreateDocumentFromConfig(...)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/datamodel/low/v3/create_document.go:26
github.com/pb33f/libopenapi.(*document).BuildV3Model(0xc0013994c0)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/document.go:316 +0xb3
github.com/speakeasy-api/openapi-generation/v2/pkg/generate.(*Generator).loadDoc(0xc0066a7548, {0x2eaebc0, 0xc00520e900}, {0xc005000000, 0x2f5de, 0x2f5df}, {0x7ffdf8d6b366, 0x1f}, 0x0, 0xc0013db900)
        /home/trist/workspace/openapi-generation/pkg/generate/generate.go:522 +0x793
github.com/speakeasy-api/openapi-generation/v2/pkg/generate.(*Generator).processSDKDocs.func1({0x7ffdf8d6b3a8, 0x2})
        /home/trist/workspace/openapi-generation/pkg/generate/generate.go:400 +0x51b
github.com/speakeasy-api/openapi-generation/v2/pkg/generate.(*Generator).processSDKDocs(0xc004fb21a0, {0x2eaebc0, 0xc00520e900}, {0xc005000000, 0x2f5de, 0x2f5df}, {0x7ffdf8d6b366, 0x1f}, 0x0, 0xc0013db900)
        /home/trist/workspace/openapi-generation/pkg/generate/generate.go:424 +0x18e
github.com/speakeasy-api/openapi-generation/v2/pkg/generate.(*Generator).generate(0xc004fb21a0, {0x2eaebc0, 0xc00520e900}, {0xc005000000, 0x2f5de, 0x2f5df}, {0x7ffdf8d6b366, 0x1f}, 0x45?)
        /home/trist/workspace/openapi-generation/pkg/generate/generate.go:40 +0xf3
github.com/speakeasy-api/openapi-generation/v2/pkg/generate.(*Generator).startGenerate(0xc00520e900?, {0x2eaebc0?, 0xc00520e900?}, {0x1a185a1, 0x8}, {0xc005000000?, 0xc00044fe70?, 0x0?}, {0x7ffdf8d6b366, 0x1f}, ...)
        /home/trist/workspace/openapi-generation/pkg/generate/generator.go:561 +0xba
github.com/speakeasy-api/openapi-generation/v2/pkg/generate.(*Generator).Generate(0xc004fb21a0, {0x2eaeb88?, 0x3a67120?}, {0xc005000000, 0x2f5de, 0x2f5df}, {0x7ffdf8d6b366, 0x1f}, {0x7ffdf8d6b3a0, 0x4}, ...)
        /home/trist/workspace/openapi-generation/pkg/generate/generator.go:391 +0x1b8
github.com/speakeasy-api/openapi-generation/v2/internal/generate.Generate({0x198?, 0x19fb440?}, {{0x7ffdf8d6b3a0, 0x4}, {0x7ffdf8d6b366, 0x1f}, {0x7ffdf8d6b389, 0x13}, {0x0, 0x0}, ...})
        /home/trist/workspace/openapi-generation/internal/generate/generate.go:91 +0xe59
main.main()
        /home/trist/workspace/openapi-generation/cmd/generate/main.go:27 +0x411

goroutine 8350 [runnable]:
github.com/pb33f/libopenapi/index.runIndexFunction.func2()
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:484
runtime.goexit()
        /home/linuxbrew/.linuxbrew/opt/go/libexec/src/runtime/asm_amd64.s:1650 +0x1
created by github.com/pb33f/libopenapi/index.runIndexFunction in goroutine 1
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:484 +0x32

goroutine 10902 [runnable]:
gopkg.in/yaml%2ev3.(*encoder).must(0xc0031fcc00?, 0x1?)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:84 +0xca
gopkg.in/yaml%2ev3.(*encoder).emit(0xc0031fcc00)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:81 +0x31
gopkg.in/yaml%2ev3.(*encoder).mappingv(0xc0031fcc00, {0x0?, 0x0}, 0xc00126e810)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:264 +0x13d
gopkg.in/yaml%2ev3.(*encoder).mapv(...)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:187
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1871fa0?, 0xc00456fe60?, 0x11fe5db?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:164 +0x7f9
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1844e80?, 0xc004957db0?, 0x98?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:162 +0x785
gopkg.in/yaml%2ev3.(*encoder).marshal.(*encoder).mapv.func1()
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:192 +0xe5
gopkg.in/yaml%2ev3.(*encoder).mappingv(0xc0031fcc00, {0x0?, 0x0}, 0xc00126ece0)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:265 +0x14a
gopkg.in/yaml%2ev3.(*encoder).mapv(...)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:187
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1871fa0?, 0xc00456fcb0?, 0x11fe5db?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:164 +0x7f9
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1844e80?, 0xc004957b20?, 0x98?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:162 +0x785
gopkg.in/yaml%2ev3.(*encoder).marshal.(*encoder).mapv.func1()
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:192 +0xe5
gopkg.in/yaml%2ev3.(*encoder).mappingv(0xc0031fcc00, {0x0?, 0x0}, 0xc00126f1b0)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:265 +0x14a
gopkg.in/yaml%2ev3.(*encoder).mapv(...)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:187
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1871fa0?, 0xc00456fc80?, 0x7ff48c73fb60?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:164 +0x7f9
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1844e80?, 0xc004957630?, 0x98?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:162 +0x785
gopkg.in/yaml%2ev3.(*encoder).marshal.(*encoder).mapv.func1()
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:192 +0xe5
gopkg.in/yaml%2ev3.(*encoder).mappingv(0xc0031fcc00, {0x0?, 0x0}, 0xc0066a3680)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:265 +0x14a
gopkg.in/yaml%2ev3.(*encoder).mapv(...)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:187
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1871fa0?, 0xc004417470?, 0x11fe5db?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:164 +0x7f9
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1844e80?, 0xc0006ee100?, 0x98?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:162 +0x785
gopkg.in/yaml%2ev3.(*encoder).marshal.(*encoder).mapv.func1()
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:192 +0xe5
gopkg.in/yaml%2ev3.(*encoder).mappingv(0xc0031fcc00, {0x0?, 0x0}, 0xc0066a3b50)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:265 +0x14a
gopkg.in/yaml%2ev3.(*encoder).mapv(...)
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:187
gopkg.in/yaml%2ev3.(*encoder).marshal(0xc0031fcc00, {0x0, 0x0}, {0x1871fa0?, 0xc004416840?, 0xc0066a3bd8?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:164 +0x7f9
gopkg.in/yaml%2ev3.(*encoder).marshalDoc(0xc0031fcc00, {0x0, 0x0}, {0x1871fa0?, 0xc004416840?, 0x1844e80?})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/encode.go:105 +0x12e
gopkg.in/yaml%2ev3.Marshal({0x1871fa0, 0xc004416840})
        /home/trist/go/pkg/mod/gopkg.in/yaml.v3@v3.0.1/yaml.go:222 +0x308
github.com/pb33f/libopenapi/utils.IsYAML({0xc0005c4000, 0x2f5de})
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/utils/utils.go:530 +0x87
github.com/pb33f/libopenapi/datamodel.ExtractSpecInfoWithDocumentCheck.func1.1()
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/datamodel/spec_info.go:105 +0x4c
created by github.com/pb33f/libopenapi/datamodel.ExtractSpecInfoWithDocumentCheck.func1 in goroutine 1
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/datamodel/spec_info.go:104 +0x23d

goroutine 8348 [runnable]:
github.com/pb33f/libopenapi/index.(*SpecIndex).GetOperationCount(0xc0062e2000)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/spec_index.go:970 +0x448
github.com/pb33f/libopenapi/index.runIndexFunction.func1(0x7?, 0xc001820330?)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:485 +0x1b
created by github.com/pb33f/libopenapi/index.runIndexFunction in goroutine 1
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:484 +0x32

goroutine 8351 [runnable]:
github.com/pb33f/libopenapi/index.runIndexFunction.func2()
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:484
runtime.goexit()
        /home/linuxbrew/.linuxbrew/opt/go/libexec/src/runtime/asm_amd64.s:1650 +0x1
created by github.com/pb33f/libopenapi/index.runIndexFunction in goroutine 1
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/utility_methods.go:484 +0x32

goroutine 10904 [chan send]:
github.com/pb33f/libopenapi/index.(*SpecIndex).MapNodes(0xc0062e2000, 0xc00591df40)
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/map_index_nodes.go:69 +0x165
created by github.com/pb33f/libopenapi/index.createNewIndex in goroutine 1
        /home/trist/go/pkg/mod/github.com/pb33f/libopenapi@v0.13.20/index/spec_index.go:71 +0x114
exit status 2
```

**Comment by TristanSpeakEasy** [comment]
Seems to be this change its breaking on https://github.com/pb33f/libopenapi/compare/82c9e21df12e613f27d7347233e61670bb59fcdf...7a5a4bcb2d32166f2cdd4a409df13b7a7d4acfe5#diff-b767c091b243f51cdff2b28c2c806ad3f4e1476a0c7dcc858a1b22f57fbfd7eeR110

**Comment by daveshanley** [comment]
Which test is causing this?

**Comment by TristanSpeakEasy** [comment]
Sorry not in a test and I haven't been able to get a minimal reproducible yet. We are seeing this with certain use cases in our product, but not all the time.

**Comment by TristanSpeakEasy** [comment]
So when it does happen, it is when we load the openapi document twice in a row, which will hopefully help me track down a reproducible

**Comment by TristanSpeakEasy** [comment]
yep okay this test reproduces it:

```
func TestDocument_LoadDocTwice(t *testing.T) {
	spec, err := os.ReadFile("test_specs/speakeasy-test.yaml")
	require.NoError(t, err)

	doc, err := NewDocumentWithConfiguration(spec, &datamodel.DocumentConfiguration{
		BasePath:                            "./test_specs",
		IgnorePolymorphicCircularReferences: true,
		IgnoreArrayCircularReferences:       true,
		AllowFileReferences:                 true,
	})
	require.NoError(t, err)

	_, errs := doc.BuildV3Model()
	require.Empty(t, errs)

	doc, err = NewDocumentWithConfiguration(spec, &datamodel.DocumentConfiguration{
		BasePath:                            "./test_specs",
		IgnorePolymorphicCircularReferences: true,
		IgnoreArrayCircularReferences:       true,
		AllowFileReferences:                 true,
	})
	require.NoError(t, err)

	_, errs = doc.BuildV3Model()
	require.Empty(t, errs)
}
```

**Comment by daveshanley** [comment]
I don't know how, but `sync.Map` is NOT threadsafe. I literally used that structure to stop this exact issue. However for some reason, it's still getting upset, when bombarded by multiple threads. 

Some RW locks will solve this.

**Comment by TristanSpeakEasy** [comment]
fixed here https://github.com/pb33f/libopenapi/pull/205/commits/2b128cb46534c444408cf6e8aae72f7472fcdec5

**Comment by daveshanley** [comment]
Fixed in `v0.14.0`
